web3/middleware/validation.py

```diff
@@ -27,21 +99,55 @@ def validate_chain_id(web3, chain_id):
         )
 
 
+@curry
+def check_extradata_length(val, length):
+    if not isinstance(val, (str, int, bytes)):
+        return val
+    result = HexBytes(val)
+    if len(result) > length:
+        raise ValidationError(
+            "The field extraData is %d bytes, but should be %d. "
+            "It is quite likely that you are connected to a POA chain. "
+            "Refer "
+            "http://web3py.readthedocs.io/en/latest/middleware.html#geth-style-proof-of-authority "
+            "for more details. The full extraData is: %r" % (
+                len(result), length, result
+            )
+        )
+    return val
+
+
 def transaction_normalizer(transaction):
     return dissoc(transaction, 'chainId')
 
 
-def validation_middleware(make_request, web3):
-    transaction_validator = apply_formatters_to_dict({
-        'chainId': validate_chain_id(web3),
-    })
+BLOCK_VALIDATORS = {
+    'extraData': check_extradata_length(length=32),
+}
 
-    transaction_sanitizer = compose(transaction_normalizer, transaction_validator)
 
-    def middleware(method, params):
-        if method in {'eth_sendTransaction', 'eth_estimateGas', 'eth_call'}:
-            post_validated_params = apply_formatter_at_index(transaction_sanitizer, 0, params)
-            return make_request(method, post_validated_params)
-        else:
-            return make_request(method, params)
-    return middleware
+block_validator = apply_formatters_to_dict(BLOCK_VALIDATORS)
+
+
+TRANSACTION_PARAMS_VALIDATORS = apply_validator_to_dict({
+    'chainId': validate_chain_id,
+})
+
+
+transaction_params_validator = apply_validator_at_index(
+    compose(transaction_normalizer, TRANSACTION_PARAMS_VALIDATORS),
```

Comment:
this is wrong. `transaction_normalizer` here is specific to `chainId`. This shouldn't be coupled with the logic for pluggable `TRANSACTION_PARAMS_VALIDATORS`